Lib/unittest/mock.py

```diff
@@ -503,11 +506,6 @@ def _mock_add_spec(self, spec, spec_set, _spec_as_instance=False,
 
         _spec_class = None
         _spec_signature = None
-        _spec_asyncs = []
```

Comment:
Hmm, I think this changes behaviour, e.g. on the following:
```
    def test_spec_async_attributes_instance(self):
        async_instance = AsyncClass()
        async_instance.async_func_attr = async_func

        mock_async_instance = Mock(async_instance)
        self.assertIsInstance(mock_async_instance.async_method, AsyncMock)
        self.assertIsInstance(mock_async_instance.async_func_attr, AsyncMock)
```

Reply:
Good catch! So I see two options to resolve this.

The first, which I will push for review, is where in addition to tracking `_spec_class` in the mock dict we also track `_spec_obj`. This passes your test and retains the efficiency of not eagerly checking asyncness of attributes that may never be accessed. It does mean we keep a reference to the spec object; I don't see an issue with that? It also still does change behavior in the much more subtle edge-case where an attribute is added to (or removed from) the spec instance _after_ the specced mock is created but before the corresponding attribute on the mock is accessed. I think this is probably an acceptable change, but open to suggestions otherwise.

If that approach is not acceptable, we can still gain the same speed benefits shown in the PR description for creating a bare `[Async]Mock()` by making a simpler optimization, to just put the collection of `_spec_asyncs` inside the `if spec is not None and not _is_list(spec):` block, which is the only case where it is relevant, instead of unconditionally doing it even when `spec` is `None` or a list. This also allows calling `dir(spec)` only once instead of twice. The downside here is that when we create a specced Mock, we still have to eagerly check all its attributes, even those that may never be used on the mock.

For the benchmark `./python -m timeit --setup="from unittest.mock import Mock; from types import CodeType" "Mock(CodeType)"` (using `CodeType` as just a random spec target that has some attributes and methods), current main branch is `388 usec per loop`, the top approach above reduces it to `281 usec per loop`, whereas the second approach has less impact, reducing it only to `350 usec per loop`.

Reply:
> It also still does change behavior in the much more subtle edge-case where an attribute is added to (or removed from) the spec instance after the specced mock is created but before the corresponding attribute on the mock is accessed. I think this is probably an acceptable change, but open to suggestions otherwise.

After thinking this through a bit more, I no longer think this behavior change is OK. The behavior of specced mocks is otherwise consistent that what matters is only the shape of the spec at the moment the mock is constructed; later changes to the spec object should not retroactively affect the mock. The preservation of `_spec_class` doesn't violate this consistency, because it is only used by identity for e.g. `__class__` and isinstance checks, lookups aren't performed against it.

So I will also add a test for this case, and push another update to the PR that uses the simpler (and somewhat less effective) optimization instead, where the `iscoroutinefunction` check remains eager for all attributes, but is only done when there's actually an object spec.

Reply:
Thanks @hauntsaninja , I agree with @carljm that we can just have spec of the object during mock instantiation and not to perform lookups during attribute access that might change the behavior.

Reply:
@hauntsaninja - is your test from the cpython repo or somewhere else? If "somewhere else", could we get it into this PR as well?

Reply:
@cjw296 I already added @hauntsaninja's test in this PR (as well as the more subtle variant where the spec object changes after Mock initialization.)

Reply:
@cjw296 The test isn't from anywhere, I just wrote it up to clearly communicate the change in behaviour that I noticed in the original version of this PR. It looks like carljm has already added a version of the test to this PR :-)

The updated code for _spec_asyncs lgtm!